File: backend/app/services/ward_service.py
"""Ward, Bed & OT module business logic (PRPs/ward-bed-ot-module-prp.md,
Phase 2). Routers (routers/wards.py) stay thin and delegate here -- same
"business logic lives in services/, router is thin" split as
services/consultation_service.py, services/patient_service.py,
services/pharmacy_service.py.

Unlike Pharmacy's `dispense` (engine mutates first, `authorize()` is called
only after the engine call returns -- REVIEW-AGENT flagged that ordering as
fragile in PRPs/pharmacy-module-prp.md Phase 3, since a token/branch
reassignment race would let the engine's mutation happen before the tenant
guard ever runs), every mutating function below resolves the affected
Ward/Room and calls `core.security.authorize` BEFORE invoking any
`ward_engine` function. `ward_engine`'s functions commit internally (see
that module's docstring), so there is no "authorize after the fact, but
before commit" middle ground here the way Pharmacy's transaction shape
allowed -- authorize-first is the only ordering that keeps a denied caller
from ever causing a state mutation.

BRANCH RESOLUTION JUDGMENT CALLS (flagged per the Phase 2 brief):

1. Admission/discharge/transfer/bed-status endpoints resolve the acted-on
   resource's branch by walking `Bed -> Ward -> branch_id` (transfer walks
   BOTH the old and new bed's ward) and pass the actual `Ward` ORM object to
   `authorize(current_user, "ward", "write"/"read", ward)`. `Ward` already
   has a real `branch_id` column, so no `_BranchScoped` stand-in is needed
   for these -- unlike Pharmacy, which needed one because `InventoryItem`
   sometimes isn't loaded (the two GET aggregate endpoints). This module's
   `authorize()` calls therefore all resolve to the SAME tenant guard in
   `core/security.py` (resource has `.branch_id`, callers who are not
   `system_admin` must match it) -- `authorize()` denies-by-default via that
   guard, and the `nurse`/`doctor`/`front_desk` `"ward"` policies registered
   in `core/security.py` exist only so a role that passes the tenant guard
   isn't then denied by "no policy registered" (same reminder every prior
   module's PRP includes).

2. OT scheduling reuses the SAME `"ward"` resource_type (not a second one)
   for its `authorize()` calls, passing the `Room` object directly --
   `Room.branch_id` is a real column, so this works the same way. An OT room
   is still a ward-module concern per the Phase 2 brief's own wording ("your
   call, document it") -- introducing a second resource_type here would only
   duplicate the exact same nurse/doctor/front_desk policy registrations for
   no behavioral difference, since the tenant guard (not the policy body)
   is what actually enforces branch isolation in every case below.

3. `list_beds`/`list_ot_schedules` (the two GET list endpoints): a
   `system_admin` caller may omit the branch filter entirely ("all
   branches"), matching `pharmacy_service.get_low_stock`/`get_expiring`'s
   established convention. Every other role MUST resolve to a concrete own-
   branch value -- for `list_beds` this is the `branch_id` query param
   (required, and rejected with 422 if it doesn't match the caller's own
   branch, exactly `pharmacy_service._resolve_branch_filter`'s mismatch
   rule); for `list_ot_schedules` there is no `branch_id` query param in the
   contract at all (only `room_id`/`surgeon_id`/`from_time`/`to_time`), so
   the caller's own branch is applied as a silent `Room.branch_id` join
   filter instead of a client-suppliable parameter -- a non-admin caller
   simply cannot see another branch's OT schedule, there is nothing to
   mismatch-reject here since there is no field carrying a conflicting
   claim.

4. Both list endpoints additionally call `authorize(..., "read", ...)` with
   a `_BranchScoped` stand-in once a concrete branch_id is known (mirroring
   Pharmacy's use of that dataclass for the same "no single ORM row" reason)
   -- defense-in-depth alongside the query-level filtering, not the only
   thing preventing cross-branch reads.

RANGE-BOUND EXTRACTION: `AdmissionResponse`/`OTScheduleResponse` need plain
`start_time`/`end_time` fields, but `Admission.stay_range`/
`OTSchedule.time_range` come back from the ORM as psycopg2 `Range` objects
(see `ward_engine._lower_bound_isoformat`'s docstring for the same
observation). `_range_bound_as_datetime` below reads `.lower`/`.upper` off
the already-loaded range and returns a `datetime` (parsing it if the driver
ever hands back a string instead, mirroring `_lower_bound_isoformat`'s own
`isinstance(..., str)` guard) -- `stay_range.upper` is `None` for an
ongoing/undischarged admission and is passed through as `None` rather than
assumed to always be set.
"""

import uuid
from dataclasses import dataclass
from datetime import datetime

from fastapi import HTTPException, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.security import authorize, get_caller_branch_id
from app.models.patient import Patient
from app.models.resource import Doctor, Room
from app.models.user import User, UserRole
from app.models.ward import Admission, Bed, BedStatus, OTSchedule, Ward
from app.schemas.ward import (
    AdmissionCreate,
    AdmissionResponse,
    BedMatrixEntryResponse,
    BedResponse,
    BedStatusUpdateRequest,
    DischargeRequest,
    OTScheduleCreate,
    OTScheduleResponse,
    TransferRequest,
)
from app.services import ward_engine


@dataclass(frozen=True)
class _BranchScoped:
    """Minimal stand-in resource for `authorize()`'s tenant guard when there
    is no single ORM row to check against -- the two GET list endpoints, when
    scoped to one branch. Same pattern as `pharmacy_service._BranchScoped`."""

    branch_id: uuid.UUID


def _range_bound_as_datetime(value: object) -> datetime | None:
    """Extract a `TSTZRANGE` bound (already-loaded psycopg2 `Range.lower` /
    `.upper`) as a `datetime`, or `None` if the bound is unset (an
    open-ended admission's `.upper`). See module docstring's "RANGE-BOUND
    EXTRACTION" section."""
    if value is None:
        return None
    if isinstance(value, datetime):
        return value
    if isinstance(value, str):
        return datetime.fromisoformat(value)
    raise TypeError(f"unexpected range bound type: {type(value)!r}")


def _shape_admission(admission: Admission) -> AdmissionResponse:
    return AdmissionResponse(
        id=admission.id,
        patient_id=admission.patient_id,
        bed_id=admission.bed_id,
        admitted_by=admission.admitted_by,
        start_time=_range_bound_as_datetime(admission.stay_range.lower),
        end_time=_range_bound_as_datetime(admission.stay_range.upper),
        discharged_at=admission.discharged_at,
    )


def _shape_ot_schedule(ot_schedule: OTSchedule) -> OTScheduleResponse:
    start_time = _range_bound_as_datetime(ot_schedule.time_range.lower)
    end_time = _range_bound_as_datetime(ot_schedule.time_range.upper)
    assert start_time is not None and end_time is not None  # OT slots are always closed ranges
    return OTScheduleResponse(
        id=ot_schedule.id,
        room_id=ot_schedule.room_id,
        patient_id=ot_schedule.patient_id,
        surgeon_id=ot_schedule.surgeon_id,
        start_time=start_time,
        end_time=end_time,
    )


def _get_bed_or_404(db: Session, bed_id: uuid.UUID) -> Bed:
    bed = db.get(Bed, bed_id)
    if bed is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "bed not found")
    return bed


def _get_ward_for_bed(db: Session, bed: Bed) -> Ward:
    ward = db.get(Ward, bed.ward_id)
    assert ward is not None  # FK guarantees this; bed couldn't exist otherwise
    return ward


def _get_patient_or_404(db: Session, patient_id: uuid.UUID) -> Patient:
    patient = db.get(Patient, patient_id)
    if patient is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "patient not found")
    return patient


def _get_admission_or_404(db: Session, admission_id: uuid.UUID) -> Admission:
    admission = db.get(Admission, admission_id)
    if admission is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "admission not found")
    return admission


# --- Admissions --------------------------------------------------------------


def admit(db: Session, current_user: User, payload: AdmissionCreate) -> AdmissionResponse:
    """POST /api/v1/wards/admissions. Validates bed/patient existence and
    authorizes against the bed's ward BEFORE calling `ward_engine.admit_patient`
    (see module docstring) -- `BedNotAvailableError`/`BedConflictError`/
    `ResourceBusyError` from the engine itself propagate uncaught to the
    router, which maps them to 409/409/423 respectively."""
    bed = _get_bed_or_404(db, payload.bed_id)
    _get_patient_or_404(db, payload.patient_id)
    ward = _get_ward_for_bed(db, bed)

    authorize(current_user, "ward", "write", ward)

    admission = ward_engine.admit_patient(
        db,
        patient_id=payload.patient_id,
        bed_id=payload.bed_id,
        start_time=payload.start_time,
        admitted_by=current_user.id,
    )
    return _shape_admission(admission)


def discharge(
    db: Session, current_user: User, admission_id: uuid.UUID, payload: DischargeRequest
) -> AdmissionResponse:
    """PATCH /api/v1/wards/admissions/{id}/discharge. Resolves the admission's
    bed's ward and authorizes BEFORE calling `ward_engine.discharge_patient`
    -- `AlreadyDischargedError`/`ResourceBusyError` from the engine propagate
    uncaught to the router (409/423)."""
    admission = _get_admission_or_404(db, admission_id)
    bed = _get_bed_or_404(db, admission.bed_id)
    ward = _get_ward_for_bed(db, bed)

    authorize(current_user, "ward", "write", ward)

    updated = ward_engine.discharge_patient(
        db,
        admission_id=admission_id,
        discharged_by=current_user.id,
        discharge_time=payload.discharge_time,
    )
    return _shape_admission(updated)


def transfer(
    db: Session, current_user: User, admission_id: uuid.UUID, payload: TransferRequest
) -> AdmissionResponse:
    """PATCH /api/v1/wards/admissions/{id}/transfer. Authorizes against BOTH
    the old and the new bed's ward before calling
    `ward_engine.transfer_patient` -- either mismatch is a 403, per the
    Phase 2 brief ("check both")."""
    admission = _get_admission_or_404(db, admission_id)
    old_bed = _get_bed_or_404(db, admission.bed_id)
    old_ward = _get_ward_for_bed(db, old_bed)
    new_bed = _get_bed_or_404(db, payload.new_bed_id)
    new_ward = _get_ward_for_bed(db, new_bed)

    authorize(current_user, "ward", "write", old_ward)
    authorize(current_user, "ward", "write", new_ward)

    new_admission = ward_engine.transfer_patient(
        db,
        admission_id=admission_id,
        new_bed_id=payload.new_bed_id,
        actor_user_id=current_user.id,
    )
    return _shape_admission(new_admission)


# --- Bed status ---------------------------------------------------------------


def update_bed_status(
    db: Session, current_user: User, bed_id: uuid.UUID, payload: BedStatusUpdateRequest
) -> BedResponse:
    """PATCH /api/v1/wards/beds/{id}/status. Authorizes against the bed's
    ward BEFORE calling `ward_engine.set_bed_status` --
    `IllegalBedStatusTransition`/`ResourceBusyError` from the engine
    propagate uncaught to the router (409/423)."""
    bed = _get_bed_or_404(db, bed_id)
    ward = _get_ward_for_bed(db, bed)

    authorize(current_user, "ward", "write", ward)

    updated_bed = ward_engine.set_bed_status(
        db,
        bed_id=bed_id,
        requested_status=BedStatus(payload.status),
        actor_user_id=current_user.id,
    )
    return BedResponse.model_validate(updated_bed)


# --- OT scheduling --------------------------------------------------------------


def schedule_ot(db: Session, current_user: User, payload: OTScheduleCreate) -> OTScheduleResponse:
    """POST /api/v1/wards/ot-schedule. Validates room/patient/surgeon
    existence BEFORE calling the engine (so a bad id gets a clean 404 rather
    than a raw FK `IntegrityError` from the engine's insert -- see the Phase
    2 brief), authorizes against the room BEFORE calling the engine (see
    module docstring point 2), then calls `ward_engine.schedule_ot`.
    `OTConflictError`/`ResourceBusyError` propagate uncaught to the router
    (409/423)."""
    room = db.get(Room, payload.room_id)
    if room is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "room not found")
    _get_patient_or_404(db, payload.patient_id)
    surgeon = db.get(Doctor, payload.surgeon_id)
    if surgeon is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "surgeon not found")

    authorize(current_user, "ward", "write", room)

    ot_schedule = ward_engine.schedule_ot(
        db,
        room_id=payload.room_id,
        patient_id=payload.patient_id,
        surgeon_id=payload.surgeon_id,
        start_time=payload.start_time,
        duration_minutes=payload.duration_minutes,
        actor_user_id=current_user.id,
    )
    return _shape_ot_schedule(ot_schedule)


# --- List / branch-scoped GETs -----------------------------------------------


def _resolve_branch_filter(current_user: User, branch_id_param: uuid.UUID | None) -> uuid.UUID | None:
    """Shared branch-resolution helper for `list_beds`. `system_admin` may
    omit `branch_id` to mean "all branches"; every other role MUST supply
    `branch_id` and it MUST match their own (token-claims-preferred, via
    `get_caller_branch_id` -- never `current_user.branch_id` directly, see
    the Phase 2 brief's explicit warning about the Pharmacy module's
    original mistake). Mismatch or omission for a non-admin caller is 422,
    mirroring `pharmacy_service._resolve_branch_filter`'s "never silently
    ignore conflicting/missing client input" discipline."""
    if current_user.role == UserRole.system_admin:
        return branch_id_param

    caller_branch_id = get_caller_branch_id(current_user)
    if caller_branch_id is None:
        raise HTTPException(
            status.HTTP_422_UNPROCESSABLE_ENTITY,
            "your account has no branch assigned; contact an administrator",
        )
    caller_branch_uuid = uuid.UUID(caller_branch_id)
    if branch_id_param is None:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, "branch_id is required")
    if branch_id_param != caller_branch_uuid:
        raise HTTPException(
            status.HTTP_422_UNPROCESSABLE_ENTITY,
            "branch_id in the request does not match your assigned branch",
        )
    return caller_branch_uuid


def list_beds(
    db: Session,
    current_user: User,
    branch_id_param: uuid.UUID | None,
    ward_id_param: uuid.UUID | None,
    status_param: BedStatus | None,
) -> list[BedMatrixEntryResponse]:
    """GET /api/v1/wards/beds?branch_id=&ward_id=&status="""
    branch_id = _resolve_branch_filter(current_user, branch_id_param)
    if branch_id is not None:
        authorize(current_user, "ward", "read", _BranchScoped(branch_id=branch_id))
    # else: system_admin querying across every branch -- no single resource to
    # authorize against; system_admin already bypasses the tenant guard inside
    # authorize() (core/security.py's _admin_bypass), and no other role can
    # reach this branch (non-admin always resolves to a concrete branch_id
    # above).

    query = select(
        Bed.id,
        Bed.ward_id,
        Ward.name,
        Ward.branch_id,
        Bed.label,
        Bed.status,
    ).join(Ward, Ward.id == Bed.ward_id)

    if branch_id is not None:
        query = query.where(Ward.branch_id == branch_id)
    if ward_id_param is not None:
        query = query.where(Bed.ward_id == ward_id_param)
    if status_param is not None:
        query = query.where(Bed.status == status_param)

    rows = db.execute(query).all()
    return [
        BedMatrixEntryResponse(
            id=row[0],
            ward_id=row[1],
            ward_name=row[2],
            branch_id=row[3],
            label=row[4],
            status=row[5].value if isinstance(row[5], BedStatus) else row[5],
        )
        for row in rows
    ]


def list_ot_schedules(
    db: Session,
    current_user: User,
    room_id_param: uuid.UUID | None,
    surgeon_id_param: uuid.UUID | None,
    from_time: datetime | None,
    to_time: datetime | None,
) -> list[OTScheduleResponse]:
    """GET /api/v1/wards/ot-schedule?room_id=&surgeon_id=&from_time=&to_time=

    At least one of `room_id`/`surgeon_id`/(`from_time` AND `to_time`) is
    required -- 422 otherwise (same "no unbounded scan" rule every prior
    module's list endpoint uses). Branch-scoped via `Room.branch_id`: unlike
    `list_beds`, there is no `branch_id` query param in this contract at all
    -- a non-admin caller's own branch is applied as a silent join filter
    instead (see module docstring point 3)."""
    if room_id_param is None and surgeon_id_param is None and not (from_time and to_time):
        raise HTTPException(
            status.HTTP_422_UNPROCESSABLE_ENTITY,
            "at least one of room_id, surgeon_id, or from_time+to_time is required",
        )

    query = select(
        OTSchedule.id,
        OTSchedule.room_id,
        OTSchedule.patient_id,
        OTSchedule.surgeon_id,
        OTSchedule.time_range,
        Room.branch_id,
    ).join(Room, Room.id == OTSchedule.room_id)

    if current_user.role == UserRole.system_admin:
        pass  # all branches
    else:
        caller_branch_id = get_caller_branch_id(current_user)
        if caller_branch_id is None:
            raise HTTPException(
                status.HTTP_422_UNPROCESSABLE_ENTITY,
                "your account has no branch assigned; contact an administrator",
            )
        caller_branch_uuid = uuid.UUID(caller_branch_id)
        authorize(current_user, "ward", "read", _BranchScoped(branch_id=caller_branch_uuid))
        query = query.where(Room.branch_id == caller_branch_uuid)

    if room_id_param is not None:
        query = query.where(OTSchedule.room_id == room_id_param)
    if surgeon_id_param is not None:
        query = query.where(OTSchedule.surgeon_id == surgeon_id_param)
    if from_time is not None and to_time is not None:
        range_literal = f"[{from_time.isoformat()},{to_time.isoformat()})"
        query = query.where(OTSchedule.time_range.op("&&")(range_literal))

    rows = db.execute(query).all()
    results = []
    for row in rows:
        time_range = row[4]
        start_time = _range_bound_as_datetime(time_range.lower)
        end_time = _range_bound_as_datetime(time_range.upper)
        assert start_time is not None and end_time is not None
        results.append(
            OTScheduleResponse(
                id=row[0],
                room_id=row[1],
                patient_id=row[2],
                surgeon_id=row[3],
                start_time=start_time,
                end_time=end_time,
            )
        )
    return results

File: frontend/src/services/wardService.ts
import axios from "axios";

import { api } from "./api";
import type { Admission, BedMatrixEntry, BedStatus, OTSchedule } from "../types";

/**
 * This file is the only place in the frontend allowed to know about the
 * backend's snake_case wire format for the Ward, Bed & OT module -- see
 * PRPs/ward-bed-ot-module-prp.md "ENDPOINTS". Every exported function
 * converts to/from the camelCase types in `types/index.ts` at the boundary,
 * same pattern as `pharmacyService.ts` / `labService.ts` /
 * `consultationService.ts`.
 *
 * `branchId` handling mirrors `pharmacyService.ts`: `GET /wards/beds` is
 * "branch-scoped like Pharmacy" per the PRP's ENDPOINTS section, so a
 * caller with their own branch (nurse/doctor/front_desk) simply never
 * sends `branch_id` (the backend derives it from the session), and only a
 * `system_admin` (whose `branchId` is `null`) supplies it explicitly.
 */

interface BedMatrixEntryWire {
  id: string;
  ward_id: string;
  ward_name: string;
  branch_id: string;
  label: string;
  status: BedStatus;
}

interface AdmissionWire {
  id: string;
  patient_id: string;
  bed_id: string;
  admitted_by: string;
  start_time: string;
  end_time: string | null;
  discharged_at: string | null;
}

interface OTScheduleWire {
  id: string;
  room_id: string;
  patient_id: string;
  surgeon_id: string;
  start_time: string;
  end_time: string;
}

/** Shape of the 200 body from `PATCH /wards/beds/{id}/status` -- a partial
 * bed, not the full `BedMatrixEntry` (no `ward_name`/`branch_id`). */
interface BedStatusWire {
  id: string;
  ward_id: string;
  label: string;
  status: BedStatus;
}

/** Shape of the 409 body from `PATCH /wards/beds/{id}/status`. */
interface BedStatusConflictWire {
  detail: string;
  current: BedStatus;
  requested: BedStatus;
}

export type OTConflictReason = "surgeon_busy_ot" | "surgeon_busy_appointment" | "room_conflict";

/** Shape of the 409 body from `POST /wards/ot-schedule`. */
interface OTConflictWire {
  detail: string;
  reason: OTConflictReason;
}

export interface GetBedMatrixParams {
  branchId?: string;
  wardId?: string;
  status?: BedStatus;
}

export interface AdmitPatientData {
  patientId: string;
  bedId: string;
  startTime: string;
}

export interface ScheduleOTData {
  roomId: string;
  patientId: string;
  surgeonId: string;
  startTime: string;
  durationMinutes: number;
}

export interface ListOTSchedulesParams {
  roomId?: string;
  surgeonId?: string;
  fromTime?: string;
  toTime?: string;
}

/** Partial bed shape returned by `updateBedStatus` -- see `BedStatusWire`. */
export interface BedStatusResult {
  id: string;
  wardId: string;
  label: string;
  status: BedStatus;
}

/**
 * Thrown on a 409 from `updateBedStatus()` -- the requested manual
 * transition isn't one of the four legal ones (see the PRP's
 * `set_bed_status` design). Carries the backend's actual `current`/
 * `requested` status strings so the UI can say e.g. "cannot go from
 * occupied to available" instead of a generic "invalid transition".
 */
export class IllegalBedStatusTransitionError extends Error {
  current: BedStatus;
  requested: BedStatus;

  constructor(message: string, current: BedStatus, requested: BedStatus) {
    super(message);
    this.name = "IllegalBedStatusTransitionError";
    this.current = current;
    this.requested = requested;
  }
}

/**
 * Thrown on a 409 from `scheduleOT()`. Carries the backend's `reason` so
 * the UI can distinguish "surgeon busy in another OT" vs. "surgeon has a
 * conflicting clinic appointment" vs. "room already booked" -- see the
 * PRP's design decision #3 (the surgeon-conflict check is app-level only,
 * spanning both `ot_schedules` and `appointments`).
 */
export class OTConflictError extends Error {
  reason: OTConflictReason;

  constructor(message: string, reason: OTConflictReason) {
    super(message);
    this.name = "OTConflictError";
    this.reason = reason;
  }
}

function toBedMatrixEntry(wire: BedMatrixEntryWire): BedMatrixEntry {
  return {
    id: wire.id,
    wardId: wire.ward_id,
    wardName: wire.ward_name,
    branchId: wire.branch_id,
    label: wire.label,
    status: wire.status,
  };
}

function toAdmission(wire: AdmissionWire): Admission {
  return {
    id: wire.id,
    patientId: wire.patient_id,
    bedId: wire.bed_id,
    admittedBy: wire.admitted_by,
    startTime: wire.start_time,
    endTime: wire.end_time,
    dischargedAt: wire.discharged_at,
  };
}

function toOTSchedule(wire: OTScheduleWire): OTSchedule {
  return {
    id: wire.id,
    roomId: wire.room_id,
    patientId: wire.patient_id,
    surgeonId: wire.surgeon_id,
    startTime: wire.start_time,
    endTime: wire.end_time,
  };
}

export async function getBedMatrix(params: GetBedMatrixParams = {}): Promise<BedMatrixEntry[]> {
  const query: Record<string, string> = {};
  if (params.branchId) {
    query.branch_id = params.branchId;
  }
  if (params.wardId) {
    query.ward_id = params.wardId;
  }
  if (params.status) {
    query.status = params.status;
  }
  const { data } = await api.get<BedMatrixEntryWire[]>("/api/v1/wards/beds", {
    params: Object.keys(query).length > 0 ? query : undefined,
  });
  return data.map(toBedMatrixEntry);
}

export async function admitPatient(data: AdmitPatientData): Promise<Admission> {
  const body = {
    patient_id: data.patientId,
    bed_id: data.bedId,
    start_time: data.startTime,
  };
  const { data: wire } = await api.post<AdmissionWire>("/api/v1/wards/admissions", body);
  return toAdmission(wire);
}

/** `dischargeTime` is optional per the PRP's ENDPOINTS table -- omitted
 * entirely from the body (not sent as `null`) when not supplied, letting
 * the backend default it to "now". */
export async function dischargePatient(admissionId: string, dischargeTime?: string): Promise<Admission> {
  const body = dischargeTime ? { discharge_time: dischargeTime } : {};
  const { data: wire } = await api.patch<AdmissionWire>(
    `/api/v1/wards/admissions/${admissionId}/discharge`,
    body,
  );
  return toAdmission(wire);
}

/** Returns the NEW admission (a new row, not an update of the old one) --
 * see the PRP's `transfer_patient` design (discharge-then-admit). */
export async function transferPatient(admissionId: string, newBedId: string): Promise<Admission> {
  const body = { new_bed_id: newBedId };
  const { data: wire } = await api.patch<AdmissionWire>(
    `/api/v1/wards/admissions/${admissionId}/transfer`,
    body,
  );
  return toAdmission(wire);
}

/**
 * Throws `IllegalBedStatusTransitionError` on a 409 -- callers should
 * `instanceof`-branch on it before falling back to `extractErrorMessage`,
 * same pattern as `pharmacyService.ts`'s `dispense()`.
 */
export async function updateBedStatus(
  bedId: string,
  status: Extract<BedStatus, "available" | "blocked">,
): Promise<BedStatusResult> {
  try {
    const { data: wire } = await api.patch<BedStatusWire>(`/api/v1/wards/beds/${bedId}/status`, {
      status,
    });
    return { id: wire.id, wardId: wire.ward_id, label: wire.label, status: wire.status };
  } catch (error) {
    if (axios.isAxiosError<BedStatusConflictWire>(error) && error.response?.status === 409) {
      const body409 = error.response.data;
      throw new IllegalBedStatusTransitionError(body409.detail, body409.current, body409.requested);
    }
    throw error;
  }
}

/**
 * Throws `OTConflictError` on a 409 -- callers should `instanceof`-branch
 * on it before falling back to `extractErrorMessage`.
 */
export async function scheduleOT(data: ScheduleOTData): Promise<OTSchedule> {
  const body = {
    room_id: data.roomId,
    patient_id: data.patientId,
    surgeon_id: data.surgeonId,
    start_time: data.startTime,
    duration_minutes: data.durationMinutes,
  };
  try {
    const { data: wire } = await api.post<OTScheduleWire>("/api/v1/wards/ot-schedule", body);
    return toOTSchedule(wire);
  } catch (error) {
    if (axios.isAxiosError<OTConflictWire>(error) && error.response?.status === 409) {
      const body409 = error.response.data;
      throw new OTConflictError(body409.detail, body409.reason);
    }
    throw error;
  }
}

export async function listOTSchedules(params: ListOTSchedulesParams = {}): Promise<OTSchedule[]> {
  const query: Record<string, string> = {};
  if (params.roomId) {
    query.room_id = params.roomId;
  }
  if (params.surgeonId) {
    query.surgeon_id = params.surgeonId;
  }
  if (params.fromTime) {
    query.from_time = params.fromTime;
  }
  if (params.toTime) {
    query.to_time = params.toTime;
  }
  const { data } = await api.get<OTScheduleWire[]>("/api/v1/wards/ot-schedule", {
    params: Object.keys(query).length > 0 ? query : undefined,
  });
  return data.map(toOTSchedule);
}
